### PR TITLE
Import hints refactoring

### LINKED ISSUE #542: Import collision shape without adding redundant static node
I'm working on a 3D space sim/4X game, and while the current flags are useful (-colonly, -col, etc.), they all add a static body by default. My base objects are already static bodies, and all I'd need is to import the collision shape, without adding anything else.
I think that godot could be improved in one of two ways, the sophisticated one and the simple one:
- Detect that the collision shape is being added to a node of a type that uses collision shapes, and abstain from adding a static body node in that case, removing redundancy and clutter. (area, static body, etc.)
- Just add another flag to signal when you want the shape to be imported without static mesh (for example -colshape)

The first option would require to check the parent type before adding static bodies and do so only when the parent is not aware of collision shapes. It's a tougher beast, because it would change the result, and so, could break people's projects if they expect the current redundant layout.

The second option would require just another flag, as the ones already present, with an even simpler implementation. It would not break any previous works, becasue it would only impact those that use the new flag.

### LINKED ISSUE #35690: Inconsistent object name for generated collision
**Godot version:**
68eae6b6e

**OS/device including version:** Arch Linux

**Issue description:** When generating collision using `-convcol` hint for glTF, `StaticBody` and `CollisionShape` objects are named in snake_case. But the editor uses CamelCase for objects on the scene by default.

**Steps to reproduce:**

1. Download model model with `-convcol` hints: [Strump.blend](https://github.com/Zoriya/asset-sources/blob/master/Objects/Vegetation/Trees/Strump.blend)
3. Open in Blender.
2. Export as glTF.
3. Import in Godot.
4. You will see `static_collision` name for `StaticBody` object and `shape<number>` for `CollisionShape` objects.

**Minimal reproduction project:**
[Inconsistent object name for generated collision.tar.gz](https://github.com/godotengine/godot/files/4127399/Inconsistent.object.name.for.generated.collision.tar.gz)

### PR DESCRIPTION
**Pull request by Shatur**
~An implementation of https://github.com/godotengine/godot-proposals/issues/542.~

~This PR improves 3D models import workflow. As described in proposal,  I've added 4 additional import hints:~

~1. `shape`~
~2. `shapeonly`~
~3. `convshape`~
~4. `convshapeonly`~

~Such import hints works like `col`, `colonly`, `convcolonly` and `convcol`, but don't create a StaticBody. See proposal for details.~

~Documentation was also updated: https://github.com/godotengine/godot-docs/pull/3622~

Now contains only refactoring.

Since the `_fix_node` function that is responsible for this is very huge and with a lot of duplicated code - I split it into several helper functions to simplify adding the new hints.

Also node name were parsed for hints in each `if`. I fixed it, now the name parsed only once.

Each object added from the import hints had a hardcoded name in the code (`static_body` for `col` hint, for example). Now the name is determined automatically based on the type of object and respects editor's case settings, closes #35690.

Also I fixed coding style (for example, sometimes variables were not in `snake_case`)

Fixes: https://github.com/godotengine/godot-proposals/issues/542

### CODE REVIEW

**Comment by fire** [comment]
Is there any consensus for removing col part from names?

**Comment by Shatur95** [comment]
@fire, I don't meant to remove `col` part from old hints. I meant to remove `col` parts from the proposed new hints that I will add: `convcolshapeonly` -> `convshapeonly`, `colshape`-> `shape`. Just for simplicity.

File: scene/main/node.h
Comment (Shatur95) [inline scene/main/node.h:49]: I need it to access `NameCasing`. Please, let me know if this is okay.

File: editor/import/resource_importer_scene.cpp
Comment (Shatur95) [inline editor/import/resource_importer_scene.cpp:273]: I added a return here because `AnimationPlayer` can't have any hints.

**Comment by Shatur95** [comment]
I tested it with the following `.blend` files:
[TestFiles.zip](https://github.com/godotengine/godot/files/4706454/TestFiles.zip)
I triggered all cases (including parenting), works as expected.

But I have a crash when opening a model with convex collision in [this](https://github.com/godotengine/godot/blob/a8787d1ae56006c00b4eecc3506dbc5ec763a8a5/core/math/quick_hull.cpp#L386) line. The issue exist in master branch (it is not related to my changes). Because of this in the attachment below you can two files: `AllImportCases.blend` (to test on 3.2) and `AllImportCasesExceptConvex.blend` (to test on master branch).

**Comment by Shatur95** [comment]
[Here](https://pastebin.com/ThKu3ML3) is the patch with the changes for 3.2 branch.

**Comment by Shatur95** [comment]
@akien-mga, sorry for bothering, but could you review it?
I see that there is a lot of PR's and I think that mine could be lost among them.

File: editor/import/resource_importer_scene.cpp
Comment (akien-mga) [inline editor/import/resource_importer_scene.cpp:271]: We don't use `auto` nor in-condition-assignments in the Godot codebase.

**Comment by akien-mga** [comment]
It's a huge patch and it's not my area of expertise, so this will have to wait sadly.

File: editor/import/resource_importer_scene.cpp
Comment (Shatur95) [reply editor/import/resource_importer_scene.cpp:271]: Hm... Before implementation I checked codebase and found `auto` usages, for example [here](https://github.com/godotengine/godot/blob/36f61030262c7661aab8407c784fd1e72eea219d/modules/gdnavigation/nav_map.cpp#L164) :(
As for me, `auto` helpes **only** in cases where type is iterator or known from the context. Here class name is already known from `cast_to`. And `auto` just helps to avoid double class typing. [clan-tidy](https://clang.llvm.org/extra/clang-tidy/checks/modernize-use-auto.html) have even a rule to use `auto` in only such cases.
But, please, let me know if you not agree with this, I will remove it of course.

This is not `in-condition-assignment`, this is an initialization inside `if` that comes from [C++17](https://en.cppreference.com/w/cpp/language/if). This feature allow to avoid polluting scope with variables that needed only inside `if` statement.
Also, let me know if you do not use it to, I will remove it too.

**Comment by Shatur95** [comment]
Not a problem, I was just afraid that the patch would be lost because some time was WIP.

File: editor/import/resource_importer_scene.cpp
Comment (akien-mga) [reply editor/import/resource_importer_scene.cpp:271]: > Before implementation I checked codebase and found auto usages, for example here :(

That's an oversight, or an exception due to the use of the very `std` overly verbose iterators that prompted the creation of `auto` in the first place. `auto` is banned in the Godot codebase, we like our types explicit.

Comment (Shatur95) [reply editor/import/resource_importer_scene.cpp:271]: Okay, I will remove it :)

Can I leave modern C++17 initialization inside `if`? It like initialization inside `for`.

Comment (akien-mga) [reply editor/import/resource_importer_scene.cpp:271]: > This is not in-condition-assignment, this is an initialization inside if that comes from C++17. This feature allow to avoid polluting scope with variables that needed only inside if statement.

We only recently started using C++17 (from a C++03 "C with classes" coding style), and we're taking a conservative approach to the use of new features.

We use the features which make obvious sense, like better templating options, guaranteed copy elision, class member initialization, etc., but for more debatable features like this one which make the code (IMO) harder to parse and more foreign to C / classical C++, I would err on the side of not using them.

It's an extremely common construct in the Godot codebase to do:
```
AnimationPlayer *ap = Object::cast_to<AnimationPlayer>(p_node);
if (ap) {

}
```

If we wanted to change this, it should be done all at once in the whole codebase (thousands of call sites) and not left to individual contributor preference.

Comment (Shatur95) [reply editor/import/resource_importer_scene.cpp:271]: Not a problem, thanks for the explanations!

The last question. In one function I need to return a `Pair`. Can I use [Structured binding declaration](https://en.cppreference.com/w/cpp/language/structured_binding)?
```cpp
// With (will no work without auto)
auto [node_name, hint] = function()
static_body->set_name(node_name);

// Without
Pair<String, Hint> fixedNode = function();
static_body->set_name(fixedNode.second);
```

Comment (akien-mga) [reply editor/import/resource_importer_scene.cpp:271]: As a rule of thumb, if a feature is C++17 only, you shouldn't use it unless it solves a use case that couldn't be done before C++17.

In this case an explicit `Pair<String, Hint>` is definitely what we **want**. We want to know the container we're using, and the types it's holding. `auto [a, b]` is definitely an anti-feature for us.

Comment (Shatur95) [reply editor/import/resource_importer_scene.cpp:271]: Okay, thanks, will fix it today!

Comment (Shatur95) [reply editor/import/resource_importer_scene.cpp]: Done!

**Comment by Shatur95** [comment]
@akien-mga, is there a chance to see it in 3.2.2 release?

**Comment by akien-mga** [comment]
Not for 3.2.2, but once merged in `master` it could be cherry-picked for a future 3.2.x release.

**Comment by Shatur95** [comment]
@akien-mga, is there something preventing this branch from being merged?

**Comment by akien-mga** [comment]
It needs a review by someone familiar with import code, to assess both the code and the usefulness of these additions. More import hints mean more code to maintain for all supported files formats (glTF, FBX, DAE, ESCN), and not all of them are currently properly supported by all formats.

**Comment by fire** [comment]
Let me see if I can put it my todo list when I get time during weekends.

**Comment by Shatur95** [comment]
> More import hints mean more code to maintain for all supported files formats (glTF, FBX, DAE, ESCN), and not all of them are currently properly supported by all formats.

No, no, sorry if I confuse you. This code works regardless of the format. It takes an already parsed model, checks if it name has import hints and just applies them. This PR also simplifies this code:
1. Name were parsed in each "if". Now the name parsing happens once.
2. Each object added from the import hints had a hardcoded name in the code (`static_body` for `col` hint, for example). Now the name is determined automatically based on the type of object and respects editor's case settings.

> Let me see if I can put it my todo list when I get time during weekends.

It would be cool, thanks!

**Comment by fire** [comment]
Because there's a large amount of combinations, can you help me test this by making tests cases like https://github.com/godotengine/godot-tests/tree/master/tests/blend_export

**Comment by Shatur95** [comment]
@fire, sure. But do you want to test all hints on all formats? `ResourceImporterScene` is format-independent.

**Comment by fire** [comment]
The idea is we create the source blend and use https://github.com/godotengine/godot-tests/blob/master/tests/blend_export/run_tests.bat to convert to the formats we want. I should add collada too.

**Comment by fire** [comment]
I need to review this, but normally for a pr to be accepted it should be merged into logical commits or one commit. You have 10.

**Comment by Shatur95** [comment]
At the moment, each commit is doing one thing. There are several commits because to add options I had to refactor the hint mechanism. But now it looks much better as to me :)
Should I squash all commits?

**Comment by fire** [comment]
Please squash the commits.

**Comment by Shatur95** [comment]
@fire, done. But the current master have issues with empty objects (they don't have `empty_draw_type` property) and QuickHull. So it's probably better to test on 3.2.

File: editor/import/resource_importer_scene.cpp
Comment (fire) [inline editor/import/resource_importer_scene.cpp:265]: Auto is banned.

Comment (Shatur95) [reply editor/import/resource_importer_scene.cpp]: Sorry, I just returned `Pair` and used `auto` to decompose the variables. Replaced by normal use of `Pair`. Maybe create a structure for it? Using `first` and `second` looks bad.

**Comment by fire** [comment]
Compiling your branch and going over the test cases. Need to check that the pr the hints. Then go over the code.

**Comment by fire** [comment]
Is it correct that the collision shapes are at origin? 
image

**Comment by Shatur95** [comment]
> Is it correct that the collision shapes are at origin?

No, sorry, just forgot to apply translation to shapes :) Fixed.

**Comment by fire** [comment]
Check also scale and rotation. The entire transform.

**Comment by Shatur95** [comment]
@fire, sure, I applied the entire transform. Just forgot about it :) Should be okay now.

**Comment by fire** [comment]
image

Is this correct?

**Comment by Shatur95** [comment]
> Is this correct?

Here is how `additional-import-hints` looks on my machine:
изображение

Have you added another custom-sized cube to this file? If so, that's ok, only the form with the same transformation is imported.

**Comment by fire** [comment]
I used the one from https://github.com/godotengine/godot-tests/.

**Comment by Shatur95** [comment]
> Look at the proposal 39105 and the fbx export.

It looks in the following way on my machine:

изображение

The only thing I did was remove the `convcol` and `convcolonly` models. Godot crashing in convex collision generation on the current master (even without this PR and it works on 3.2).
Convex collision generation don't crash on your machine?

> Since it works in dae, gltf2 and obj, it's not caused by this pr.

Yes, I think so.

> I noticed obj set to scene was broken.

The same here even on 3.2 :(

File: editor/import/resource_importer_scene.cpp
Comment (fire) [inline editor/import/resource_importer_scene.cpp:325]: Is this correct?

You replace the pointer by the static body. I haven't looked into this enough.

Comment (fire) [reply editor/import/resource_importer_scene.cpp:325]: I tend to use queue delete, but the definition of next engine tick before saving is a bit confusing.

Comment (Shatur95) [reply editor/import/resource_importer_scene.cpp:325]: Not sure, I'm new here, but this is how it was https://github.com/godotengine/godot/blob/59fdfc77e8f3d455ff6c86c73c27b04b022843c8/editor/import/resource_importer_scene.cpp#L397. I just added a new condition for shapes, removed redundant `set_name` and reworked parsing hint (parsing was done on every check!).

**Review by fire — APPROVED** [review]
I can't see anything obviously wrong and the tests file from blender seem correct. FBX is broken, but that's for @RevoluPowered to look at.

File: editor/import/resource_importer_scene.cpp
Comment (fire) [inline editor/import/resource_importer_scene.cpp:1053]: Is Pair from Godot? Godot tends to use structs for this.

File: editor/import/resource_importer_scene.cpp
Comment (fire) [inline editor/import/resource_importer_scene.cpp:49]: How do you use cctype? Is there a string function that can be reused?

File: scene/main/node.h
Comment (fire) [reply scene/main/node.h:48]: Ok.

File: editor/import/resource_importer_scene.cpp
Comment (fire) [reply editor/import/resource_importer_scene.cpp:1053]: If it's not a godot data type, anything else in godot will need to import C++ std library etc.

Comment (fire) [reply editor/import/resource_importer_scene.cpp:1053]: It's from core/templates/pair.h

Comment (Shatur95) [reply editor/import/resource_importer_scene.cpp:1053]: Yes :)

File: editor/import/resource_importer_scene.cpp
Comment (Shatur95) [reply editor/import/resource_importer_scene.cpp:49]: This needed for [isalpha](https://en.cppreference.com/w/cpp/string/byte/isalpha). I looked in the ustring class but didn't find an alternative. I can provide it by myself as it was done in [this](https://github.com/godotengine/godot/pull/43563/files) PR.

File: editor/import/resource_importer_scene.cpp
Comment (Shatur95) [reply editor/import/resource_importer_scene.cpp:1053]: > Godot tends to use structs for this.

Should I replace it with a struct? I also think that struct will look nicer.

File: editor/import/resource_importer_scene.cpp
Comment (fire) [reply editor/import/resource_importer_scene.cpp:49]: I don't know if doing a regex search for "a-z" and "A-Z" is more readable or not.

Comment (fire) [reply editor/import/resource_importer_scene.cpp:49]: > In locales other than "C", an alphabetic character is a character for which std::isupper() or std::islower() returns non-zero or any other character considered alphabetic by the locale.

This is also complex.

Comment (Shatur95) [reply editor/import/resource_importer_scene.cpp]: > I don't know if doing a regex search for "a-z" and "A-Z" is more readable or not.

This will not work correctly locales other than english. Is it okay? String represents a Unicode characters.

I also made a mistake. We should use [iswalpha](https://en.cppreference.com/w/cpp/string/wide/iswalpha) instead of `isalpha` for this case. But let me know if I should use something else.

**Comment by Shatur95** [comment]
@fire, I've made the latest changes in individual commits to make it easier for you to review (you can see what has changed since you last viewed). Let me know when you're done, I'll squash them.

**Comment by fire** [comment]
I changed the gltf2 import, but wanted to let you know so you can retest.

**Comment by Shatur95** [comment]
@fire, thanks! Updated to the latest master and tested - everything works.
Also convex collision now works on the latest master. But an empty objects are still not.

Could this PR be merged?

File: editor/import/resource_importer_scene.cpp
Comment (fire) [inline editor/import/resource_importer_scene.cpp:51]: What is this used for? Other than trying to not use c++17 stl, the pr looks acceptable.

**Review by fire — APPROVED** [review]

**Comment by fire** [comment]
My testing strategy was using the blender cases we made, previously one setup wasn't implemented. Happy Holidays.

File: editor/import/resource_importer_scene.cpp
Comment (Shatur95) [reply editor/import/resource_importer_scene.cpp]: I used if for [iswalpha](https://en.cppreference.com/w/cpp/string/wide/iswalpha). But you're right, it doesn't really fit here. I replaced it with `is_alpha` (that does a simple check), which I added to `ustring.h` next to `is_symbol`. Is it looks ok now?

**Comment by Shatur95** [comment]
@fire, thank you and happy holidays too!
Just solved recent small conflicts with the latest master caused by #44183.

@akien-mga, since it is approved, can it be merged?

**Comment by fire** [comment]
Conflicts still exist.

**Comment by Shatur95** [comment]
@fire, should be ok now.

**Comment by Shatur95** [comment]
@akien-mga, sorry for ping, but is there something preventing this branch from being merged?

**Comment by akien-mga** [comment]
I'm not familiar enough with the 3D assets pipeline to know if adding those additional hints is the best way to support the requested feature.

I tried to get @reduz to review it but he didn't seem to understand the use case yet. I would suggest discussing it with him on IRC.

**Comment by reduz** [comment]
I don't see this as used often enough to justify the extra complexity in the import system (and there hasn't been much enthusiasm for the proposal in general during all this time). I am also on the process of rewriting the asset import system for 4.0, which will better support general import plugins. Hopefully this functionality can be added as an import plugin at the time.

On my side, I don't agree with merging with this PR and suggest waiting for the new import system to be available to re-evaluate.

**Comment by Shatur95** [comment]
@reduz, I completely understand your point of view about complexity. But current collision import hints (`-col`, `-colonly`, `-convcol` and `-convcolonly`) are not useful. The creation of an additional `StaticBody` is never needed. Because of this I made new additional import hints.
But If you disagree with the idea of adding new import hints, how about replacing the behavior of current collision import hints `-col`, `-colonly`, `-convcol` and `-convcolonly`) with the new one from this PR (but only for 4.0, of course)? The new imports hints from this pull request are much more useful and can fully replace the old one.

**Comment by Shatur95** [comment]
@reduz, or are you planning to remove import hints entirely in favor of a new import system?

**Comment by reduz** [comment]
Its not that I agree or not, its simply that most other users seem to be doing things a different way than you and don't seem to need this. The reason why I am against merging it is simply that it's complex and no one seems to care about it, whether you or I believe it's useful or not is irrelevant. Again I suggest waiting for the new import system, which will give you more control on this, and proper plugins so you can implement this as an add-on.

**Comment by Shatur95** [comment]
@reduz, got it!

I removed newly added hints. So, now the PR only contains the following:

1. Refactoring (I moved duplicate code and complex parts into separate functions that only do one thing) with renaming (now names consistent with other engine parts, for example `p_` is used for function arguments).
2. Parsing speedup. Now node name parsed only once into enum instead of calling [parsing function every time](https://github.com/godotengine/godot/blob/a7baf013524ccd271bd123d8aef93a73b450a888/editor/import/resource_importer_scene.cpp#L373).
3. Remove extra `set_name`, fixes #35690.

@akien-mga, can this be merged?

**Comment by Shatur95** [comment]
As discussed on IRC, this code in this PR can be affected by the implementation of https://github.com/godotengine/godot-proposals/issues/1823. So, I need to wait for the new import system and only then apply these changes if necessary.

**Comment by Shatur95** [comment]
I applied all fixes in a separate PR's except parsing speedup. I'm going to think again about how this can be done more elegantly and will send a PR soon.

**Comment by CsloudX** [comment]
??? Any New information for this?

**Pull request closed without merging**
